src/main/java/org/prebid/server/bidder/grid/GridBidder.java

```diff
@@ -1,56 +1,243 @@
+                        .headers(HttpUtil.headers())
+                        .payload(modifiedRequest)
+                        .body(mapper.encode(modifiedRequest))
+                        .build();
+
+        return Result.of(Collections.singletonList(httpRequest), errors);
+    }
+
+    private List<Imp> modifyImps(List<Imp> imps, List<BidderError> errors) {
+        final List<Imp> modifiedImps = new ArrayList<>();
+        for (Imp imp : imps) {
+            try {
+                final ExtImpGrid extImpGrid = mapper.mapper().convertValue(imp.getExt(), ExtImpGrid.class);
+                validateImpExt(extImpGrid, imp.getId());
+                modifiedImps.add(modifyImp(imp, extImpGrid));
+            } catch (IllegalArgumentException | PreBidException e) {
+                errors.add(BidderError.badInput(e.getMessage()));
+            }
+        }
+
+        return modifiedImps;
+    }
+
+    private static void validateImpExt(ExtImpGrid extImpGrid, String impId) {
+        final ExtImpGridBidder extImpGridBidder = extImpGrid != null ? extImpGrid.getBidder() : null;
+        final Integer uid = extImpGridBidder != null ? extImpGridBidder.getUid() : null;
+        if (uid == null || uid == 0) {
+            throw new PreBidException(String.format("Empty uid in imp with id: %s", impId));
         }
+    }
 
-        final GridExtImpData extImpData = gridExtImp != null ? gridExtImp.getData() : null;
-        final GridExtImpDataAdServer adServer = extImpData != null ? extImpData.getAdServer() : null;
+    private Imp modifyImp(Imp imp, ExtImpGrid extImpGrid) {
+        final ExtImpGridData extImpData = extImpGrid.getData();
+        final ExtImpGridDataAdServer adServer = extImpData != null ? extImpData.getAdServer() : null;
         final String adSlot = adServer != null ? adServer.getAdSlot() : null;
-        if (StringUtils.isNotEmpty(adSlot)) {
 
-            final GridExtImp modifiedGridExtImp = gridExtImp.toBuilder()
+        if (StringUtils.isNotEmpty(adSlot)) {
+            final ExtImpGrid modifiedExtImpGrid = extImpGrid.toBuilder()
                     .gpid(adSlot)
                     .build();
+            return imp.toBuilder()
+                    .ext(mapper.mapper().valueToTree(modifiedExtImpGrid))
+                    .build();
+        }
+        return imp;
+    }
 
-            return imp.toBuilder().ext(mapper.mapper().valueToTree(modifiedGridExtImp)).build();
+    private Keywords getKeywordsFromImpExt(JsonNode extImp) {
+        try {
+            return mapper.mapper().convertValue(extImp, GRID_EXT_TYPE_REFERENCE).getBidder().getKeywords();
+        } catch (IllegalArgumentException e) {
+            return null;
         }
+    }
 
-        return imp;
+    private BidRequest modifyRequest(BidRequest bidRequest, Keywords firstImpKeywords, List<Imp> imp) {
+        final User user = bidRequest.getUser();
+        final String userKeywords = user != null ? user.getKeywords() : null;
```

Comment:
Consider to use `org.prebid.server.util.ObjectUtil#getIfNotNull`
Same for `siteKeywords`.